macros/src/main/scala/com/softwaremill/macwire/internals/CompanionCrimper.scala

```diff
@@ -0,0 +1,46 @@
+package com.softwaremill.macwire.internals
+
+import scala.reflect.macros.blackbox
+
+private[macwire] class CompanionCrimper [C <: blackbox.Context, T: C#WeakTypeTag](val c: C, log: Logger) {
+  import c.universe._
+
+  lazy val dependencyResolver = new DependencyResolver[c.type](c, log)
+
+  lazy val targetType: Type = implicitly[c.WeakTypeTag[T]].tpe
+
+  lazy val companionType: Option[Type] = if(targetType.companion == NoType) None else Some(targetType.companion)
+
+  def isCompanionApply(method: Symbol): Boolean =
+    method.isMethod &&
+    method.isPublic &&
+    method.asMethod.returnType <:< targetType &&
+    method.asMethod.name.decodedName.toString == "apply"
+
+  lazy val applies: Option[List[Symbol]] = log.withBlock("Looking for apply methods of Companion Object") {
+    val as: Option[List[Symbol]] = companionType.map(_.members.filter(isCompanionApply).toList)
+    as.foreach(x => log.withBlock(s"There are ${x.size} apply methods:" ) { x.foreach(c => log(showApply(c))) })
+    as
+  }
+
+  lazy val apply: Option[Symbol] = applies.flatMap( _ match {
+    case applyMethod :: Nil => Some(applyMethod)
+    case _ => None
+  })
+
+  lazy val applySelect: Option[Select] = apply.map(a => Select(Ident(targetType.typeSymbol.companion), a))
```

Comment:
Consider this usage of `lazy val`s:

```scala
object MacwireMacros { ...
  def wire_impl[T: c.WeakTypeTag](c: blackbox.Context): c.Expr[T] = {
    import c.universe._

    val constructorCrimper = new ConstructorCrimper[c.type, T](c, log)
    val companionCrimper = new CompanionCrimper[c.type, T](c, log)

    lazy val targetType = companionCrimper.targetType.toString
    lazy val whatWasWrong: String = {
      if(constructorCrimper.constructor.isEmpty && companionCrimper.companionType.isEmpty)
        s"Cannot find a public constructor nor a companion object for [$targetType]"
      else if(companionCrimper.applies.isDefined && companionCrimper.applies.get.isEmpty)
        s"Companion object for [$targetType] has no apply methods constructing target type."
      else if(companionCrimper.applies.isDefined && companionCrimper.applies.get.size > 1)
        s"No public primary constructor found for $targetType and multiple matching apply methods in its companion object were found."
      else s"Target type not supported for wiring: $targetType. Please file a bug report with your use-case."
    }

    val code: Tree = (constructorCrimper.constructorTree orElse companionCrimper.applyTree) getOrElse
      c.abort(c.enclosingPosition, whatWasWrong)
    log(s"Generated code: ${showCode(code)}, ${showRaw(code)}")
    c.Expr(code)
  }
```

`companionCrimper.applyTree` is a `lazy val`. That means that if `constructorCrimper.constructorTree` is `Some` it won't be computed - faster compilation time.

You may ask then why the `constructorCrimper.constructorTree` is lazy? It seems that it is computed always. Well, not exactly. In `Macros.wire_impl` - yes it is. But `MacwireAkkaMacros` uses (indirectly) `ConstructorCrimper` as well but not this particular lazy val. If we made `constructorCrimper.constructorTree` eager `val` it would be computed even if we don't used it.

Ok. So do we really need to have all `val`s in this [Crimper](https://www.google.pl/search?q=crimper&source=lnms&tbm=isch&sa=X&ved=0ahUKEwiTrt_kke3RAhXIbZoKHQjcAx4Q_AUICCgB&biw=1920&bih=1070)-things to be lazy? Actually no, we don't. However it's easier to make them all lazy than manually track what is needed and when. Can we have some of them private and some not? Yes - but sometimes it's easier to type `companionCrimper. [ctrl+space]` and with IDE support see what's inside and pick up what is needed. 

Btw. Lazy vals aren't publicly exposed. They are visible only in macwire` package (see `private[macwire] class ConstructorCrimper ...`).

Does it make any sense now?